### ISSUE TITLE
scaler-september-open-source-Kunal-Garg-12

### ISSUE THREAD

**Issue by Kunal-Garg-12** [scaleracademy/scaler-september-open-source-challenge#268]
name: Kunal Garg
github_user_name: Kunal-Garg-12

**Comment by Kunal-Garg-12** [comment]
day 2 --> Forked and cloned the repository

- URL of forked repository: https://github.com/Kunal-Garg-12/scaler-september-open-source-challenge-fork
- Screenshot of the cloned project in your system 
![Screenshot (1773)](https://user-images.githubusercontent.com/63821392/188085360-19270f3b-d6bf-42a4-b297-4d5be535f334.png)

**Comment by Kunal-Garg-12** [comment]
day 2 --> Forked and cloned the repository

- URL of forked repository: https://github.com/Kunal-Garg-12/scaler-september-open-source-challenge-fork
- Screenshot of the cloned project on the system
![repo](https://user-images.githubusercontent.com/63821392/188086306-717f94cf-25ed-4733-b0da-ddee5e5e52da.jpg)

**Comment by Kunal-Garg-12** [comment]
day 3 --> Created a branch in my forked repository

Screenshot showing the list of all the current branches inside of the forked clone repository
![branch](https://user-images.githubusercontent.com/63821392/188267674-bb2c31be-7ed4-471c-ad28-79e810831fa3.jpg)

**Comment by Kunal-Garg-12** [comment]
day 4 --> Updated Fork Repo From the Original Repo

- Used "git remote add upstream https://github.com/scaleracademy/scaler-september-open-source-challenge.git" to add the original repo as remote repo. Then used "git pull origin main" to update local repo
![d4 1](https://user-images.githubusercontent.com/63821392/188309977-adda9796-e065-4d15-b341-30e574a3c399.jpg)

- Used "git push origin main" to update my forked repo on GitHub
![d4 2](https://user-images.githubusercontent.com/63821392/188310000-b5a4da30-ce05-4c98-bf25-92371c93040f.jpg)

- Used "git push origin main:Kunal-Garg-12-details" to update my branch on GitHub
![d4 3](https://user-images.githubusercontent.com/63821392/188310030-5f1c0554-6739-4562-9e9a-9bc0e199ef16.jpg)

**Comment by Kunal-Garg-12** [comment]
day 5--> Created a markdown file in contributors/Kunal-Garg-12/ and submitted a pull request on the main challenge repo.
![d5](https://user-images.githubusercontent.com/63821392/188441540-84624e51-b592-4440-a0d3-d090a087945f.jpg)

**Comment by Kunal-Garg-12** [comment]
day 6--> Updated pull request created in Challenge 5 (day 5)
![d6 pull request updated](https://user-images.githubusercontent.com/63821392/188587987-af6a10e4-dd53-4dd0-9582-f2f69c7591fc.jpg)
![d6 pull request updated 2](https://user-images.githubusercontent.com/63821392/188588003-4f2c7eef-6a81-45cb-978f-c235bf705610.jpg)

**Comment by Kunal-Garg-12** [comment]
day 6--> Corrected mistakes by resetting main and redoing challenge 5 and challenge 6 via username-details branch instead of main branch and deleted the original pull request
URL for new PR: https://github.com/scaleracademy/scaler-september-open-source-challenge/pull/937
Original PR: 
![d6 deleted PR from main branch](https://user-images.githubusercontent.com/63821392/188687954-2298c94e-0674-4502-85b1-59d6a26992ff.jpg)

New PR:
![d6 made corrections and using branch](https://user-images.githubusercontent.com/63821392/188687766-7128d50d-e877-4141-9969-fced4dd44c24.jpg)

**Comment by Kunal-Garg-12** [comment]
day7--> Squashed the last two commits

**Comment by Kunal-Garg-12** [comment]
day7--> Squashed the last two commits
![d7 Squashed commits](https://user-images.githubusercontent.com/63821392/188915627-55bc886e-f485-4adc-878e-b0e4a000b93a.jpg)

**Comment by Kunal-Garg-12** [comment]
day 8--> Created username-2.md file and pushed it to remote repo. Verified that the change appears in the Pull request.
![d8 create usernam-2 md and push](https://user-images.githubusercontent.com/63821392/189163383-c89d80cf-64c4-43f9-8a0b-cb3572d0310f.jpg)

**Comment by Kunal-Garg-12** [comment]
day 9--> Used git reset to remove last commit from PR
![d9 git reset](https://user-images.githubusercontent.com/63821392/189369017-090441fb-a2aa-4dce-b5f0-f49eac132661.jpg)

**Comment by Kunal-Garg-12** [comment]
day 10-->Setup the ability to sign commits locally in your systems and updated username.md and pushed a signed commit for the same
![d10 signed commit](https://user-images.githubusercontent.com/63821392/189483047-77966523-9bd4-4394-b3aa-db5c0d68f188.jpg)

**Comment by Kunal-Garg-12** [comment]
day 10-->Setup the ability to sign commits locally on the system and updated username.md and pushed a signed commit for the same
![d10 signed commit](https://user-images.githubusercontent.com/63821392/189483084-09198797-e075-406d-b30e-c7729c37a888.jpg)

**Comment by Kunal-Garg-12** [comment]
day 11--> Created a new branch from username-details branch created in previous challenge
![d11 create new branch](https://user-images.githubusercontent.com/63821392/189528926-536e36c9-de8e-4192-a150-da765cbff16a.jpg)
![d11 all branches](https://user-images.githubusercontent.com/63821392/189528929-42d17fc5-51fd-443e-839d-9b23887b717e.jpg)

**Comment by Kunal-Garg-12** [comment]
day 12--> Verified if my PR has no more than 2 commits after squashing of commits till day 7.

**Comment by Kunal-Garg-12** [comment]
day 13--> Learnt about the git stash command and practiced it.
![d13 git stash](https://user-images.githubusercontent.com/63821392/189942960-68f8c316-6451-498c-b882-76584e7d5b13.jpg)

**Comment by Kunal-Garg-12** [comment]
day 14--> Created a case of merge conflict in locally and try to resolved
![d14 1 merge conflict](https://user-images.githubusercontent.com/63821392/190216490-3c8b9d83-f7c5-474f-bf91-535b2ffc7d74.jpg)
![d14 2](https://user-images.githubusercontent.com/63821392/190216502-11fee801-940b-4fc6-95d8-b30c936b6438.jpg)
![d14 3](https://user-images.githubusercontent.com/63821392/190216512-86783627-6fe4-4691-a5d2-2eec33eac991.jpg)
 it

**Comment by Kunal-Garg-12** [comment]
day 14--> Created a case of merge conflict in locally and try to resolved it
![d14](https://user-images.githubusercontent.com/63821392/190216816-f2043b81-bf33-4e2a-a793-d1912223f916.png)

**Comment by Kunal-Garg-12** [comment]
day 15--> Updated forked repo and username-details branch. Added name in challengers-list.md file and pushed change.
![d15 Updated repo-branch and added name in challengers-list](https://user-images.githubusercontent.com/63821392/190454619-49007150-e047-4554-90f4-594801fb04fe.jpg)

**Comment by Kunal-Garg-12** [comment]
day 16--> Resolved merge conflict using local Git and VS Code for challengers-list.md file
![d16](https://user-images.githubusercontent.com/63821392/190707816-83b9a679-7a4d-4451-b4df-3ce631f0696a.png)
![d16 PR GitHub](https://user-images.githubusercontent.com/63821392/190707823-ad53a20d-d91e-4732-876f-abda84e58341.jpg)

**Comment by Kunal-Garg-12** [comment]
day 17--> Learnt about GitHub Desktop. Tried out creating, editing, branching, pulling and pushing on GitHub Desktop.
Link to practice repo: https://github.com/Kunal-Garg-12/GitHub-Desktop-Practice
![d17 Practiced GitHub Desktop](https://user-images.githubusercontent.com/63821392/190866624-f1222807-54d6-4554-8e8b-98a8ce2ba06f.jpg)

**Comment by Kunal-Garg-12** [comment]
day 18--> Learnt about gists and markdown. Created two gists and pushed them.
![d18 gists](https://user-images.githubusercontent.com/63821392/190923923-408f943e-8682-4f6f-8dd5-7b3677a8531c.jpg)

**Comment by Kunal-Garg-12** [comment]
day 19--> Learnt about linting. Added a sample linted code.  Linted: gist-solutions.md to pass checks
![d19 Linting](https://user-images.githubusercontent.com/63821392/191041576-cdf8c26b-3ed6-47c0-a32a-4713ab541c62.jpg)

**Comment by Kunal-Garg-12** [comment]
day20 --> Created a project. Made a TODO list. Shifted tasks to in progress and done
![d20 1 create a project board](https://user-images.githubusercontent.com/63821392/191321027-0d460fd5-0bc1-4a4d-aee7-c787ac8c418d.jpg)
![d20 2 in progress](https://user-images.githubusercontent.com/63821392/191321036-0a9053ee-2c8b-4add-8073-0bee0c82329a.jpg)
![d20 3 done](https://user-images.githubusercontent.com/63821392/191321039-0af58568-797c-4538-b057-a784b00f582f.jpg)

**Comment by Kunal-Garg-12** [comment]
day 21--> Created and used aliases in git
![Uploading d21 alias.jpg…]()

**Comment by Kunal-Garg-12** [comment]
day 21--> Created and used aliases in git
![d21 alias](https://user-images.githubusercontent.com/63821392/191560505-fa635f8f-cea6-4956-be1b-21836e860973.jpg)

**Comment by kuraakhilesh8230** [comment]
Congrats you have successfully completed milestone 2 and you are good to move onto next milestone ✅️
All the best😊

**Comment by Kunal-Garg-12** [comment]
day 22--> Hosted a website on GitHub and added link to it in username.md file created previously
![d22 added link to hosted page](https://user-images.githubusercontent.com/63821392/191808714-47a68e53-2fdc-4ea4-b11d-cb60d7e2d082.jpg)

**Comment by Kunal-Garg-12** [comment]
day 24--> How to  contribute to real open source projects. 

A brief on some best practices:
Always read the guidelines on contributing to a project before starting to contribute to it. This is best practice. This helps keep the project well-structured and well-formed.

Branching
The Git branching model is powerful, but we need to manage the branches correctly and effectively. Use well-defined and descriptive naming for branches but also keep the names short especially for long-lived branches. 
Master branch should always have stable version of project while development and QA should be done in separate branches.

Issues
To save time and be efficient, we should use labels and filters to search for issues we might want to take up and resolve. To make changes we would like to see in the project, we should check for existing issues related to that change and if such an issue does not exist, we should create one with appropriate format.

Pull Requests
Always make a descriptive pull request with appropriate titles. Add comments in PR to explain about why, what and how is that PR useful. Also, when committing some changes to the PR use descriptive commit messages. Signing commits is a good practice. Self-review your PR.

**Comment by Kunal-Garg-12** [comment]
day 25--> Looked on GitHub and other websites to find suitable issues I could actually contribute to but didn't find any for now. Will continue looking tomorrow and possibly find one that I can actually solve.

**Comment by Kunal-Garg-12** [comment]
day 25--> Found a repository I would like to contribute to. Asked the Issue raiser to share some more details so I can start working on it. Waiting for response.

Link to repo: https://github.com/BabaMalik/Coding-Ninjas-Introduction-to-Python
Link to issue: https://github.com/BabaMalik/Coding-Ninjas-Introduction-to-Python/issues/7

**Comment by Kunal-Garg-12** [comment]
day 26--> Didn't receive any reply from the issue raiser. Raised separate issue which I found in the repo. Fixed the issues and created a pull request. Waiting for PR to get approved.
![d26 created a PR](https://user-images.githubusercontent.com/63821392/192324304-49f09fb5-16c0-47fe-bbb5-ac6584536a50.jpg)

Looking forward to participating in Hacktoberfest

**Comment by kuraakhilesh8230** [comment]
congrats you have completed milestone-3 successfully and you are good to go🎉

**Comment by Kunal-Garg-12** [comment]
day 27--> Update: The PR got approved and merged into main. Now looking for more repos I can contribute to.
![d27 approved PR](https://user-images.githubusercontent.com/63821392/192575261-bf1db7b7-6e3d-4a92-b682-ff8eeb6767f1.jpg)


Advice for Hacktoberfest:
- You can register anytime between September 26 and October 31. So make sure you register and start contributing.
- You can register as contributor or repository manager or both if you want. Make sure to fill the registration form correctly regarding this.
- Search GitHub for label "Hacktoberfest" to find valid contributions towards Hacktoberfest.
- Do not spam PR as this will get you removed from Hacktoberfest.
- Try to make quality and impactful contributions.
- Follow contribution guidelines for every repo you contribute towards.
- You only need 4 accepted pull/merge requests to claim rewards. So again, try to make meaningful contributions instead of spamming.
- You can also make non-code contributions for example by correcting documentation.
- Join the Hacktoberfest discord.

**Comment by Kunal-Garg-12** [comment]
day 28--> As updated earlier, my PR for the issue I created was accepted. The other issue I wanted to work upon required elaboration from the issue raiser who still hasn't responded. Currently looking for more repos I can contribute to and also looking repos for Hacktoberfest that I can contribute to.

**Comment by Kunal-Garg-12** [comment]
day 29--> Suggested an idea
![d29 idea](https://user-images.githubusercontent.com/63821392/193089851-6540d77b-ce8b-4710-be6d-d18e1e37af32.jpg)

**Comment by Kunal-Garg-12** [comment]
day 30--> It is the last day of this challenge and I have learnt a lot throughout the whole challenge. I don't have any pending tasks as far as I know. Thanks to scaler team for this challenge, it gave a really good roadmap with hands on to learn about git and GitHub. Now I can finally use these technologies and make my life much easier. Thanks again!

**Comment by kuraakhilesh8230** [comment]
congrats you have completed milestone-4 successfully and you are good to go🎉
